hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/protocol/datatransfer/sasl/SaslDataTransferClient.java

```diff
@@ -405,6 +405,11 @@ public String getTargetQOP() {
     return targetQOP;
   }
 
+  @VisibleForTesting
+  public void setTargetQOP(String targetQOP) {
```

Comment:
I am not sure if this is the best way to achieve the fixes to the test case. Making a public setter for a field public only to enable a change into a unitTest breaks the abstraction and encapsulation of the Object.
Is there any other way to make this public setter?

Reply:
I totally agree with you.
Let me find a better way.

Reply:
Sounds great! Does a mock or a child class do the trick? or is it a lot of code changes to achieve?

Reply:
Sorry, but as commented in https://github.com/apache/hadoop/pull/2598#discussion_r554448244, I decided not to reset those states. But I was considering to use Java Reflection API as follows and it worked.
```
  private static void resetTargetQOP(List<DataNode> dns)
      throws NoSuchFieldException, IllegalAccessException {
    Field f = SaslDataTransferClient.class.getDeclaredField("targetQOP");
    f.setAccessible(true);
    for (DataNode dn: dns) {
      f.set(dn.getSaslClient(), null);
    }
  }
```
We can do the same thing with mockito's FieldSetter class, but I could not find it in the latest mockito version, and so I did not choose it.